# Classification work in the course-Data imaging

# Intro

In this dataset we have data of hands’ postures collected by system called Leap-Motion.
During the experiment there are two people are sitting in front of each other, both people were told to do the following: 
1. Move their hands one next to another spontaneously.
2. move their hands in an attempt to be in sync.

There are 3 modes to the experiment:<br>
**1.** **'Alone' mode**-  Each participant, in her turn, is asked to move his hand freely over the Leap-Motion controller.  <br>
**2.** **'Autonomous' mode**- Same as the 'Alone' mode, but now the two participants can see each others movement during the experiment.  <br>
**3.** **'Synchronization' mode**- The two participants are asked to moved in synchronization. They are invited
to interpret synchronization as they understand.<br>

In this notebook, I will take the data and organize it with pandas, and try to build an optimal model that given the hands’ postures I can understand in what mode the experiment was.

01:39 מה הטענה שעומדת מאחורי זה



# <b>Loading libraries</b>


In [1]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from matplotlib.image import imread
import seaborn as sns
from sklearn.model_selection import train_test_split #to perform the splitting


# <b>Loading dataset</b>

As It has already explained in the attached Word file, <br>
In each of the participants, if there is case with directories which have 2 recordings there — than the last collection should be taken.


In [2]:
#Evyatar_spontan = pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\Training\Evyatar Cohen\Evyatar636771050011691836Spontan.csv")
Evyatar_spontan = pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\Training\Evyatar Cohen\Evyatar636771052727603804Spontan.csv")
#Evyatar_sync = pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\Training\Evyatar Cohen\Evyatar636771050902934708Sync.csv")
Evyatar_sync = pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\Training\Evyatar Cohen\Evyatar636771053639929594Sync.csv")
#Evyatar_alone = pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\Training\Evyatar Cohen\Evyatar636771051823980826Alone.csv")
Evyatar_alone = pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\Training\Evyatar Cohen\Evyatar636771054555711409Alone.csv")

#Nofar_spontan = pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\Training\Nofar Social_Nuero\Nofar636759792183061005Spontan.csv")
Nofar_spontan = pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\Training\Nofar Social_Nuero\Nofar636759795182793299Spontan.csv")
#Nofar_sync = pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\Training\Nofar Social_Nuero\Nofar636759794200223541Sync.csv")
Nofar_sync = pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\Training\Nofar Social_Nuero\Nofar636759797397919664Sync.csv")
#Nofar_alone = pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\Training\Nofar Social_Nuero\Nofar636759793193137655Alone.csv")
Nofar_alone = pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\Training\Nofar Social_Nuero\Nofar636759796290435160Alone.csv")

#Oriya_spontan = pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\Training\Oriya Social_Nuero\Oriya636759799659227614Spontan.csv")
Oriya_spontan = pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\Training\Oriya Social_Nuero\Oriya636759804404113837Spontan.csv")
#Oriya_sync = pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\Training\Oriya Social_Nuero\Oriya636759803164759865Sync.csv")
Oriya_sync = pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\Training\Oriya Social_Nuero\Oriya636759806131350399Sync.csv")
#Oriya_alone = pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\Training\Oriya Social_Nuero\Oriya636759800632839612Alone.csv")
Oriya_alone = pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\Training\Oriya Social_Nuero\Oriya636759805268396661Alone.csv")

#Orya_spontan = pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\Training\Orya Kalmanovitz\OryaB636771080972009775Spontan.csv")
Orya_spontan = pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\Training\Orya Kalmanovitz\OryaB636771083605535985Spontan.csv")
Orya_sync = pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\Training\Orya Kalmanovitz\OryaB636771084494502229Sync.csv")
#Orya_sync = pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\Training\Orya Kalmanovitz\OryaB636771081904052624Sync.csv")
Orya_alone = pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\Training\Orya Kalmanovitz\OryaB636771082736601528Alone.csv")

#Revital_spontan = pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\Training\Revital Marbel\Revital636770953887274261Spontan.csv")
Revital_spontan = pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\Training\Revital Marbel\Revital636770957038873193Spontan.csv")
#Revital_sync = pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\Training\Revital Marbel\Revital636770954952844556Sync.csv")
Revital_sync = pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\Training\Revital Marbel\Revital636770957894370062Sync.csv")
Revital_alone = pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\Training\Revital Marbel\Revital636770955999435879Alone.csv")

Shahar_spontan = pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\Training\Shahar Terner\Sachar636771045591258071Spontan.csv")
Shahar_sync = pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\Training\Shahar Terner\Sachar636771046527112972Sync.csv")
Shahar_alone = pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\Training\Shahar Terner\Sachar636771047408504303Alone.csv")

#Shelly_spontan = pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\Training\Shelly Social_Nuero\Shelly636759784020992272Spontan.csv")
Shelly_spontan = pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\Training\Shelly Social_Nuero\Shelly636759787508979097Spontan.csv")
#Shelly_sync = pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\Training\Shelly Social_Nuero\Shelly636759786470414550Sync.csv")
Shelly_sync = pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\Training\Shelly Social_Nuero\Shelly636759789526313406Sync.csv")
#Shelly_alone = pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\Training\Shelly Social_Nuero\Shelly636759785254377607Alone.csv")
Shelly_alone = pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\Training\Shelly Social_Nuero\Shelly636759788383205274Alone.csv")

Yael_spontan = pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\Training\Yael Hagai\Yael636771059544754823Spontan.csv")
Yael_sync = pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\Training\Yael Hagai\Yael636771060699529085Sync.csv")
Yael_alone = pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\Training\Yael Hagai\Yael636771061659067649Alone.csv")

#Yoel_spontan = pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\Training\Yoel Raz\Yoel636760918326775654Spontan.csv")
Yoel_spontan = pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\Training\Yoel Raz\Yoel636760921663075813Spontan.csv")
#Yoel_sync = pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\Training\Yoel Raz\Yoel636760920526441748Sync.csv")
Yoel_sync = pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\Training\Yoel Raz\Yoel636760923490742912Sync.csv")
#Yoel_alone = pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\Training\Yoel Raz\Yoel636760919215510154Alone.csv")
Yoel_alone = pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\Training\Yoel Raz\Yoel636760922573364655Alone.csv")


HandRight=pd.read_csv(r"C:\Users\yuval\Hand Positions\extraFiles\HandRight.csv")


As It has already explained in the attached Word file, <br>
I need to delete the first 7 seconds in every df.

In [3]:
Evyatar_spontan

Time   Frame ID  Hand Type   # hands   Position X   Position Y  \
0     335.9984     128556      right         2    128.63570     224.0294   
1     335.9984     128556       left         2    -93.24845     143.9423   
2     336.0150     128558      right         2    129.15550     221.9923   
3     336.0150     128558       left         2    -94.86554     148.5542   
4     336.0315     128560      right         2    129.87070     220.7944   
...        ...        ...        ...       ...          ...          ...   
9049  410.9461     137193       left         2    -97.82056     193.5205   
9050  410.9625     137195      right         2    157.14880     252.6001   
9051  410.9625     137195       left         2    -99.81297     196.2407   
9052  410.9792     137197      right         2    156.92750     245.0109   
9053  410.9792     137197       left         2   -102.09200     198.4801   

       Position Z   Velocity X   Velocity Y   Velocity Z  ...       Yaw  \
0       -12.29588    -18.26243   -140.54450   -117.57790  ... -1.588251   
1       -37.81597   -109.09820    240.38980    -67.62959  ...  0.873458   
2       -15.00181     47.36464    -64.25022   -168.85850  ... -1.528183   
3       -39.00158    -80.99012    278.60220    -66.79356  ...  0.879703   
4       -18.38681     35.05422    -84.74242   -205.72030  ... -1.461710   
...           ...          ...          ...          ...  ...       ...   
9049     31.93343    -85.41756    173.52750    132.65640  ...  1.243927   
9050    -24.63861    -57.16140   -494.41450     27.38884  ... -0.041292   
9051     34.41874   -117.87250    134.38010    126.73710  ...  1.251931   
9052    -23.51586    -12.01428   -417.86870     53.70547  ... -0.043465   
9053     36.82147   -143.44210    162.22250    138.81830  ...  1.257915   

       Wrist Pos X   Wrist Pos Y   Wrist Pos Z   Elbow pos X   Elbow Pos Y  \
0         177.5125      173.5658    -14.569630      276.8753     -33.68264   
1        -147.0455      109.9380     -5.326688     -329.8577     -34.41571   
2         177.7894      171.1952    -14.677850      276.0645     -33.15913   
3        -148.2022      113.0909     -7.375025     -326.9279     -39.97873   
4         178.3730      169.9159    -15.246240      276.1767     -30.72506   
...            ...           ...           ...           ...           ...   
9049     -159.9450      167.8123     46.467390     -349.1904      13.02137   
9050      172.9846      222.5917     35.086170      300.6257      34.79204   
9051     -161.6141      169.2442     48.278710     -351.2047      14.82206   
9052      173.0201      217.8849     37.341740      300.7716      28.81221   
9053     -163.8422      171.0227     50.173370     -354.1954      17.57833   

       Elbow Pos Z   Grab Strenth   Grab Angle   Pinch Strength  
0         91.61740            0.0     1.250576              0.0  
1         93.88171            0.0     0.014736              0.0  
2         97.88171            0.0     1.069930              0.0  
3         85.98000            0.0     0.000000              0.0  
4        104.13920            0.0     0.916969              0.0  
...            ...            ...          ...              ...  
9049      21.30005            0.0     0.539401              0.0  
9050     129.14170            0.0     0.482085              0.0  
9051      23.44588            0.0     0.666431              0.0  
9052     128.65720            0.0     0.546007              0.0  
9053      25.11470            0.0     0.729553              0.0  

[9054 rows x 22 columns]

In [4]:
Evyatar_spontan = Evyatar_spontan.iloc[847:]
Evyatar_spontan


Time   Frame ID  Hand Type   # hands   Position X   Position Y  \
847   343.0061     129364       left         2   -102.28700     198.3166   
848   343.0227     129365      right         2    151.83340     204.4899   
849   343.0227     129365       left         2   -103.87210     201.4944   
850   343.0392     129367      right         2    151.40560     203.4883   
851   343.0392     129367       left         2   -106.36910     207.4667   
...        ...        ...        ...       ...          ...          ...   
9049  410.9461     137193       left         2    -97.82056     193.5205   
9050  410.9625     137195      right         2    157.14880     252.6001   
9051  410.9625     137195       left         2    -99.81297     196.2407   
9052  410.9792     137197      right         2    156.92750     245.0109   
9053  410.9792     137197       left         2   -102.09200     198.4801   

       Position Z   Velocity X   Velocity Y   Velocity Z  ...       Yaw  \
847     -2.122641   -174.57210    379.14680    28.479910  ...  1.481006   
848    -39.084380    -54.05849    -21.54555    19.405560  ... -0.560823   
849     -2.063620   -178.50250    357.85480     6.646493  ...  1.486744   
850    -38.142800    -14.11075    -56.44812    68.092860  ... -0.554623   
851     -2.337110   -126.78780    318.52840   -12.872290  ...  1.496526   
...           ...          ...          ...          ...  ...       ...   
9049    31.933430    -85.41756    173.52750   132.656400  ...  1.243927   
9050   -24.638610    -57.16140   -494.41450    27.388840  ... -0.041292   
9051    34.418740   -117.87250    134.38010   126.737100  ...  1.251931   
9052   -23.515860    -12.01428   -417.86870    53.705470  ... -0.043465   
9053    36.821470   -143.44210    162.22250   138.818300  ...  1.257915   

       Wrist Pos X   Wrist Pos Y   Wrist Pos Z   Elbow pos X   Elbow Pos Y  \
847      -164.0542      164.5314     -1.686393     -322.2086     -30.71549   
848       190.8670      170.1528      9.191355      323.1941      20.16283   
849      -165.7898      168.1620     -2.054844     -318.4708     -31.61653   
850       190.4547      169.7302     10.512530      324.1130      19.09524   
851      -168.4939      174.6985     -3.134404     -316.4541     -28.81729   
...            ...           ...           ...           ...           ...   
9049     -159.9450      167.8123     46.467390     -349.1904      13.02137   
9050      172.9846      222.5917     35.086170      300.6257      34.79204   
9051     -161.6141      169.2442     48.278710     -351.2047      14.82206   
9052      173.0201      217.8849     37.341740      300.7716      28.81221   
9053     -163.8422      171.0227     50.173370     -354.1954      17.57833   

       Elbow Pos Z   Grab Strenth   Grab Angle   Pinch Strength  
847      11.754400            0.0     1.870325         0.000000  
848     161.858900            1.0     2.704073         0.293775  
849       7.529989            0.0     1.851572         0.000000  
850     161.374600            1.0     2.422549         0.175940  
851      -1.155143            0.0     1.865896         0.000000  
...            ...            ...          ...              ...  
9049     21.300050            0.0     0.539401         0.000000  
9050    129.141700            0.0     0.482085         0.000000  
9051     23.445880            0.0     0.666431         0.000000  
9052    128.657200            0.0     0.546007         0.000000  
9053     25.114700            0.0     0.729553         0.000000  

[8207 rows x 22 columns]

In [5]:
#Let's do it for every df in the list:
Evyatar_sync= Evyatar_sync.iloc[847:]
Evyatar_alone=Evyatar_alone.iloc[847:]

Nofar_spontan=Nofar_spontan.iloc[847:]
Nofar_sync=Nofar_sync.iloc[847:]
Nofar_alone=Nofar_alone.iloc[847:]

Oriya_spontan=Oriya_spontan.iloc[847:]
Oriya_sync=Oriya_sync.iloc[847:]
Oriya_alone=Oriya_alone.iloc[847:]

Orya_spontan=Orya_spontan.iloc[847:]
Orya_sync=Orya_sync.iloc[847:]
Orya_alone =Orya_alone.iloc[847:]

Revital_spontan =Revital_spontan.iloc[847:]
Revital_sync =Revital_sync.iloc[847:]
Revital_alone =Revital_alone.iloc[847:]

Shahar_spontan =Shahar_spontan.iloc[847:] 
Shahar_sync =Shahar_sync .iloc[847:]
Shahar_alone =Shahar_alone.iloc[847:]

Shelly_spontan =Shelly_spontan .iloc[847:]
Shelly_sync =Shelly_sync .iloc[847:]
Shelly_alone =Shelly_alone .iloc[847:]

Yael_spontan =Yael_spontan .iloc[847:]
Yael_sync =Yael_sync .iloc[847:]
Yael_alone =Yael_alone .iloc[847:]

Yoel_spontan =Yoel_spontan.iloc[847:]
Yoel_sync =Yoel_sync .iloc[847:]
Yoel_alone =Yoel_alone .iloc[847:]


HandRight=HandRight.iloc[422:]


# Merge all Sync dataframes
And define the label 'Sync' as 0.

In [6]:
framesSync = [Evyatar_sync, Nofar_sync,Oriya_sync,Orya_sync,Revital_sync,Shahar_sync,Shelly_sync,Yael_sync,Yoel_sync]
dfSync= pd.concat(framesSync)
dfSync=dfSync.reset_index()
dfSync.drop(['index'],axis='columns', inplace=True)
#I will define the whole column 'HandPosition_Label' as 0 (Sync position)         
dfSync['HandPosition_Label']=0
dfSync

Time   Frame ID  Hand Type   # hands   Position X   Position Y  \
0      434.2387     139878       left         2    -80.92312     284.2069   
1      434.2553     139880      right         2    121.61940     226.1235   
2      434.2553     139880       left         2    -79.71033     282.4789   
3      434.2718     139882      right         2    121.03880     222.2049   
4      434.2718     139882       left         2    -78.84495     279.8847   
...         ...        ...        ...       ...          ...          ...   
68391  856.2631      67309      right         2    101.91230     212.0264   
68392  856.2958      67311       left         2    -81.57359     183.5054   
68393  856.2958      67311      right         2    103.84750     214.7484   
68394  856.3295      67314       left         2    -84.05124     187.8450   
68395  856.3295      67314      right         2    106.11880     218.0190   

        Position Z   Velocity X   Velocity Y   Velocity Z  ...   Wrist Pos X  \
0        29.011070     84.47700    -45.59256    134.20850  ...     -107.3702   
1        64.109630    -32.32643   -248.08270    109.73410  ...      143.3853   
2        31.691100     55.75132   -119.51820    158.45370  ...     -106.8187   
3        65.984120    -28.86916   -210.96420    109.70620  ...      143.3656   
4        35.125760     29.94153   -131.54950    221.05070  ...     -106.3970   
...            ...          ...          ...          ...  ...           ...   
68391    42.535680     79.95810    113.41030    190.77140  ...      131.8590   
68392    -7.185198    -63.24731    121.75270     62.95970  ...     -109.7250   
68393    47.070820     61.11773     95.13531    155.74530  ...      134.0416   
68394    -5.172269    -57.17249    109.20360     58.48146  ...     -109.5167   
68395    53.588680     25.46704     72.26683    190.16360  ...      137.0708   

        Wrist Pos Y   Wrist Pos Z   Elbow pos X   Elbow Pos Y   Elbow Pos Z  \
0          221.7536      46.12994     -254.8560      22.47225      40.89467   
1          166.9691      94.34527      293.7114     -20.13464     156.68940   
2          220.1253      48.09507     -258.4003      23.92692      43.43317   
3          163.3614      96.39383      296.6688     -21.13594     159.25370   
4          217.4241      50.26434     -262.3120      24.76191      42.24427   
...             ...           ...           ...           ...           ...   
68391      149.6264      66.09853      283.6185     -58.03491      95.29868   
68392      130.7078      34.03360     -295.5957     -14.97817     140.02630   
68393      151.5924      68.17693      288.9688     -54.13820      94.23959   
68394      131.6489      33.32945     -282.9649     -31.35462     135.07610   
68395      154.2158      71.46925      289.6567     -53.67185      93.98731   

        Grab Strenth   Grab Angle   Pinch Strength  HandPosition_Label  
0                0.0     0.000000              0.0                   0  
1                0.0     0.001488              0.0                   0  
2                0.0     0.000000              0.0                   0  
3                0.0     0.000000              0.0                   0  
4                0.0     0.000000              0.0                   0  
...              ...          ...              ...                 ...  
68391            0.0     0.000000              0.0                   0  
68392            0.0     0.000000              0.0                   0  
68393            0.0     0.000000              0.0                   0  
68394            0.0     0.000000              0.0                   0  
68395            0.0     0.000000              0.0                   0  

[68396 rows x 23 columns]

# Merge all Spontan dataframes
And define the label 'Spontan' as 1.

In [7]:
framesSpontan = [Evyatar_spontan,Nofar_spontan,Oriya_spontan,Orya_spontan,Revital_spontan,Shahar_spontan,Shelly_spontan,Yael_spontan,Yoel_spontan]
dfSpontan= pd.concat(framesSpontan)
dfSpontan=dfSpontan.reset_index()
dfSpontan.drop(['index'],axis='columns', inplace=True)
#I will define the whole column 'HandPosition_Label' as 0 (Sync position)         
dfSpontan['HandPosition_Label']=1
dfSpontan

Time   Frame ID  Hand Type   # hands   Position X   Position Y  \
0      343.0061     129364       left         2   -102.28700     198.3166   
1      343.0227     129365      right         2    151.83340     204.4899   
2      343.0227     129365       left         2   -103.87210     201.4944   
3      343.0392     129367      right         2    151.40560     203.4883   
4      343.0392     129367       left         2   -106.36910     207.4667   
...         ...        ...        ...       ...          ...          ...   
68493  673.4894      52512       left         2    -81.66566     203.9826   
68494  673.5233      52516      right         2    158.51870     172.6876   
68495  673.5233      52516       left         2    -77.57621     198.9182   
68496  673.5560      52519      right         2    159.35720     175.7603   
68497  673.5560      52519       left         2    -76.10054     195.9422   

        Position Z   Velocity X   Velocity Y   Velocity Z  ...   Wrist Pos X  \
0        -2.122641   -174.57210    379.14680    28.479910  ...     -164.0542   
1       -39.084380    -54.05849    -21.54555    19.405560  ...      190.8670   
2        -2.063620   -178.50250    357.85480     6.646493  ...     -165.7898   
3       -38.142800    -14.11075    -56.44812    68.092860  ...      190.4547   
4        -2.337110   -126.78780    318.52840   -12.872290  ...     -168.4939   
...            ...          ...          ...          ...  ...           ...   
68493    24.040280    240.80950   -160.56200  -222.444200  ...     -121.2401   
68494   -29.329130     62.95145     93.99306   -63.022420  ...      200.4651   
68495    20.579730    127.00290   -218.65710   -80.403090  ...     -120.1531   
68496   -30.162820     10.30027    118.67660    -7.947631  ...      199.5587   
68497    18.301860     54.47641    -89.67952   -82.426190  ...     -119.7428   

        Wrist Pos Y   Wrist Pos Z   Elbow pos X   Elbow Pos Y   Elbow Pos Z  \
0          164.5314     -1.686393     -322.2086     -30.71549     11.754400   
1          170.1528      9.191355      323.1941      20.16283    161.858900   
2          168.1620     -2.054844     -318.4708     -31.61653      7.529989   
3          169.7302     10.512530      324.1130      19.09524    161.374600   
4          174.6985     -3.134404     -316.4541     -28.81729     -1.155143   
...             ...           ...           ...           ...           ...   
68493      146.1889     44.914080     -263.3239     -65.38571     89.689400   
68494      148.9561     25.427530      400.5782      31.01108    139.279600   
68495      144.5244     44.439020     -276.8819     -56.07000     90.445170   
68496      149.5821     24.801710      390.8217      24.22654    145.633700   
68497      143.0875     43.652130     -284.9290     -46.36212    104.629500   

        Grab Strenth   Grab Angle   Pinch Strength  HandPosition_Label  
0                0.0     1.870325         0.000000                   1  
1                1.0     2.704073         0.293775                   1  
2                0.0     1.851572         0.000000                   1  
3                1.0     2.422549         0.175940                   1  
4                0.0     1.865896         0.000000                   1  
...              ...          ...              ...                 ...  
68493            0.0     1.299514         0.000000                   1  
68494            0.0     0.448368         0.000000                   1  
68495            0.0     0.542512         0.000000                   1  
68496            0.0     0.513919         0.000000                   1  
68497            0.0     0.301000         0.000000                   1  

[68498 rows x 23 columns]

# Merge all Sync&Spontan dataframes

In [8]:
#Merge all Sync and Spontan dataframes into one, with concat method as we learned previous semester

frames = [dfSpontan,dfSync]

dfSync_Spontan = pd.concat(frames)
dfSync_Spontan=dfSync_Spontan.reset_index()
dfSync_Spontan.drop(['index'],axis='columns', inplace=True)
dfSync_Spontan

Time   Frame ID  Hand Type   # hands   Position X   Position Y  \
0       343.0061     129364       left         2   -102.28700     198.3166   
1       343.0227     129365      right         2    151.83340     204.4899   
2       343.0227     129365       left         2   -103.87210     201.4944   
3       343.0392     129367      right         2    151.40560     203.4883   
4       343.0392     129367       left         2   -106.36910     207.4667   
...          ...        ...        ...       ...          ...          ...   
136889  856.2631      67309      right         2    101.91230     212.0264   
136890  856.2958      67311       left         2    -81.57359     183.5054   
136891  856.2958      67311      right         2    103.84750     214.7484   
136892  856.3295      67314       left         2    -84.05124     187.8450   
136893  856.3295      67314      right         2    106.11880     218.0190   

         Position Z   Velocity X   Velocity Y   Velocity Z  ...   Wrist Pos X  \
0         -2.122641   -174.57210    379.14680    28.479910  ...     -164.0542   
1        -39.084380    -54.05849    -21.54555    19.405560  ...      190.8670   
2         -2.063620   -178.50250    357.85480     6.646493  ...     -165.7898   
3        -38.142800    -14.11075    -56.44812    68.092860  ...      190.4547   
4         -2.337110   -126.78780    318.52840   -12.872290  ...     -168.4939   
...             ...          ...          ...          ...  ...           ...   
136889    42.535680     79.95810    113.41030   190.771400  ...      131.8590   
136890    -7.185198    -63.24731    121.75270    62.959700  ...     -109.7250   
136891    47.070820     61.11773     95.13531   155.745300  ...      134.0416   
136892    -5.172269    -57.17249    109.20360    58.481460  ...     -109.5167   
136893    53.588680     25.46704     72.26683   190.163600  ...      137.0708   

         Wrist Pos Y   Wrist Pos Z   Elbow pos X   Elbow Pos Y   Elbow Pos Z  \
0           164.5314     -1.686393     -322.2086     -30.71549     11.754400   
1           170.1528      9.191355      323.1941      20.16283    161.858900   
2           168.1620     -2.054844     -318.4708     -31.61653      7.529989   
3           169.7302     10.512530      324.1130      19.09524    161.374600   
4           174.6985     -3.134404     -316.4541     -28.81729     -1.155143   
...              ...           ...           ...           ...           ...   
136889      149.6264     66.098530      283.6185     -58.03491     95.298680   
136890      130.7078     34.033600     -295.5957     -14.97817    140.026300   
136891      151.5924     68.176930      288.9688     -54.13820     94.239590   
136892      131.6489     33.329450     -282.9649     -31.35462    135.076100   
136893      154.2158     71.469250      289.6567     -53.67185     93.987310   

         Grab Strenth   Grab Angle   Pinch Strength  HandPosition_Label  
0                 0.0     1.870325         0.000000                   1  
1                 1.0     2.704073         0.293775                   1  
2                 0.0     1.851572         0.000000                   1  
3                 1.0     2.422549         0.175940                   1  
4                 0.0     1.865896         0.000000                   1  
...               ...          ...              ...                 ...  
136889            0.0     0.000000         0.000000                   0  
136890            0.0     0.000000         0.000000                   0  
136891            0.0     0.000000         0.000000                   0  
136892            0.0     0.000000         0.000000                   0  
136893            0.0     0.000000         0.000000                   0  

[136894 rows x 23 columns]

# Merge Alone dataframes into one
And define label as 2.

In [9]:
#Merge all Alone dataframes into one, with concat method as we learned previous semester

frames_alone=[Evyatar_alone,Nofar_alone,Oriya_alone,Orya_alone,Revital_alone,Shahar_alone,Shelly_alone,Yael_alone,Yoel_alone]
dfAlone = pd.concat(frames_alone)
dfAlone=dfAlone.reset_index()
dfAlone.drop(['index'],axis='columns', inplace=True)
dfAlone

Time   Frame ID  Hand Type   # hands   Position X   Position Y  \
0      532.8434     151238       left         1    -30.37894     225.4433   
1      532.8600     151240       left         1    -32.18705     229.7671   
2      532.8773     151242       left         1    -34.76137     235.3203   
3      532.8931     151244       left         1    -37.73934     237.4320   
4      532.9097     151246       left         1    -42.22663     240.0867   
...         ...        ...        ...       ...          ...          ...   
31194  764.4578      59984       left         1   -125.81370     214.8894   
31195  764.4913      59986       left         1   -125.92320     215.6769   
31196  764.5253      59988       left         1   -125.62170     215.7428   
31197  764.5587      59989       left         1   -125.13520     215.6353   
31198  764.5913      59991       left         1   -123.99810     215.1558   

        Position Z   Velocity X   Velocity Y   Velocity Z  ...       Yaw  \
0         46.16312   -83.614320   271.978200    26.200460  ...  1.558521   
1         46.52704  -118.260000   230.771500    11.701220  ...  1.539598   
2         46.89627  -130.418000   328.530400    13.750730  ...  1.509229   
3         47.03260  -186.356500    53.857760    10.121690  ...  1.473503   
4         47.74528  -222.202000   190.160200    40.711090  ...  1.455768   
...            ...          ...          ...          ...  ...       ...   
31194     20.59324    20.170870    19.997410    37.082170  ...  0.409534   
31195     22.09264    -6.867068    21.957800    42.485630  ...  0.399021   
31196     23.39750    17.454730    -3.217659    31.446690  ...  0.379251   
31197     23.80248    25.567410    -5.648659    21.284490  ...  0.368866   
31198     24.26945    30.369090   -15.233250     6.734551  ...  0.373039   

        Wrist Pos X   Wrist Pos Y   Wrist Pos Z   Elbow pos X   Elbow Pos Y  \
0         -87.51802      183.6657      37.15565     -301.8589      54.37961   
1         -87.13968      184.9756      38.34803     -293.8652      44.03124   
2         -88.71401      189.0738      40.00905     -286.1353      35.82117   
3         -91.78994      191.0690      41.64631     -285.3000      33.80478   
4         -96.47658      193.8687      43.08831     -284.2405      30.73557   
...             ...           ...           ...           ...           ...   
31194    -146.63140      148.2971      44.40766     -258.0887     -77.52971   
31195    -146.11130      148.8045      45.68930     -253.6847     -80.13625   
31196    -144.81740      148.4588      46.66075     -249.6531     -83.33176   
31197    -143.84820      148.1024      46.72861     -248.2768     -84.26936   
31198    -142.68390      147.4703      46.75579     -247.4121     -84.85361   

        Elbow Pos Z   Grab Strenth   Grab Angle   Pinch Strength  
0          2.226311       0.000000     0.032902         0.000000  
1          2.624075       0.000000     0.136581         0.000000  
2          2.387099       0.000000     0.306545         0.000000  
3          0.430297       0.000000     0.527000         0.000000  
4         -1.736447       0.000000     0.869023         0.000000  
...             ...            ...          ...              ...  
31194    112.970600       0.116068     1.901876         0.000000  
31195    109.996700       0.134497     1.952747         0.000000  
31196    105.004300       0.120680     1.984825         0.000000  
31197    103.468300       0.097735     1.988376         0.000000  
31198    103.138000       0.065015     1.989704         0.016533  

[31199 rows x 22 columns]

# Merge Alone df with Hand Right df

First, let's check the size of dfAlone and HandRight

In [10]:
HandRight.shape

(4390, 22)

In [11]:
dfAlone.shape

(31199, 22)

In [12]:
print("As It is possible to see, after each row of left hand in dfAlone, we should put right hand from HandRight df.")
print("Therefore, we need to add more",len(dfAlone.index)-len(HandRight.index), "rows to HandRight df ")

As It is possible to see, after each row of left hand in dfAlone, we should put right hand from HandRight df.
Therefore, we need to add more 26809 rows to HandRight df 


## Turn both dataframes to the same size

In [13]:
#I will choose randomlymore 4390 out of 4390 - Allow a random selection of the same row more than once (by setting replace=True):

dfAddHandRight = HandRight.sample(n=4390,replace=True)

#and merge it with the HandRight df that already exist

framesHandToAdd=[HandRight,dfAddHandRight]

dfHandRight1 = pd.concat(framesHandToAdd)
dfHandRight1=dfHandRight1.reset_index()
dfHandRight1.drop(['index'],axis='columns', inplace=True)
dfHandRight1

Time   Frame ID  Hand Type   # hands   Position X   Position Y  \
0     135.3194      16289      right         1    11.191900     195.3075   
1     135.3357      16291      right         1    10.533080     189.0297   
2     135.3524      16293      right         1    10.243000     181.9816   
3     135.3690      16295      right         1    10.020610     174.8592   
4     135.3857      16297      right         1    10.120270     167.8396   
...        ...        ...        ...       ...          ...          ...   
8775  176.1560      20997      right         1  -174.978200     136.8923   
8776  201.9634      23969      right         1     6.871065     116.5108   
8777  164.9256      19702      right         1     0.705596     103.3047   
8778  148.6145      17821      right         1    36.423500     329.2992   
8779  190.4842      22647      right         1     1.073903     129.5454   

       Position Z   Velocity X   Velocity Y   Velocity Z  ...       Yaw  \
0        3.661716   -43.587440   -309.83740    -77.18515  ... -0.904311   
1        2.829388   -30.403600   -363.00330    -43.79815  ... -0.901862   
2        2.064330   -15.424000   -423.21200    -43.99948  ... -0.901268   
3        1.228467    -8.733602   -413.45130    -53.88054  ... -0.897777   
4        0.319165    10.939650   -389.70650    -63.49560  ... -0.892533   
...           ...          ...          ...          ...  ...       ...   
8775    18.241730   270.850000    -18.94347   -242.06540  ... -0.428341   
8776    61.474290    69.895260    -21.47004   -715.06570  ...  0.036256   
8777    33.578470  -106.743300    497.22880     47.42017  ...  0.014130   
8778    56.478440    66.776590    -12.69832     21.37463  ... -0.048613   
8779     8.511810   -19.266390    257.99120    -41.37813  ... -0.078393   

       Wrist Pos X   Wrist Pos Y   Wrist Pos Z   Elbow pos X   Elbow Pos Y  \
0        74.461410      202.4562      44.28679     282.62780     140.97120   
1        73.600600      197.4387      43.54082     280.63840     142.54010   
2        73.195370      191.7202      42.73155     279.73570     148.13820   
3        72.686710      186.1613      41.98006     278.68900     152.52150   
4        72.376450      180.5758      41.30936     277.88440     155.16360   
...            ...           ...           ...           ...           ...   
8775   -137.179700      118.0921      82.56219     -15.76801      91.38301   
8776           NaN           NaN           NaN           NaN           NaN   
8777      9.464501      109.5527     110.25140      72.03844      61.98678   
8778     51.132220      297.5986     125.29840     164.86100     127.92360   
8779     17.068810      109.6325      80.87971     118.26680      27.64994   

       Elbow Pos Z   Grab Strenth   Grab Angle   Pinch Strength  
0         203.6773       0.000000     0.626321              0.0  
1         206.7526       0.000000     0.636596              0.0  
2         209.9330       0.000000     0.701644              0.0  
3         212.1112       0.000000     0.786548              0.0  
4         213.4511       0.000000     0.816478              0.0  
...            ...            ...          ...              ...  
8775      325.1931       0.000235     1.071210              0.0  
8776           NaN       0.000000     0.823178              0.0  
8777      373.9070       0.000000     1.232087              0.0  
8778      308.9033       0.000000     0.326906              0.0  
8779      319.4344       0.000000     0.763391              0.0  

[8780 rows x 22 columns]

In [14]:
#We need to achive 31199 rows, therefore I will do it again. 

dfAddHandRight = HandRight.sample(n=4390,replace=True)

#and merge it with the dfHandRight1 df that already exist

framesHandToAdd=[dfHandRight1,dfAddHandRight]

dfHandRight2 = pd.concat(framesHandToAdd)
dfHandRight2=dfHandRight2.reset_index()
dfHandRight2.drop(['index'],axis='columns', inplace=True)
dfHandRight2


Time   Frame ID  Hand Type   # hands   Position X   Position Y  \
0      135.3194      16289      right         1     11.19190     195.3075   
1      135.3357      16291      right         1     10.53308     189.0297   
2      135.3524      16293      right         1     10.24300     181.9816   
3      135.3690      16295      right         1     10.02061     174.8592   
4      135.3857      16297      right         1     10.12027     167.8396   
...         ...        ...        ...       ...          ...          ...   
13165  191.7664      22795      right         1    -21.43827     272.0085   
13166  166.9758      19939      right         1     24.80047     167.8530   
13167  205.1294      24334      right         1     64.35109     157.7304   
13168  191.1999      22730      right         1     74.60522     267.6537   
13169  143.0998      17186      right         1     43.46407     244.6142   

        Position Z   Velocity X   Velocity Y   Velocity Z  ...       Yaw  \
0         3.661716   -43.587440  -309.837400    -77.18515  ... -0.904311   
1         2.829388   -30.403600  -363.003300    -43.79815  ... -0.901862   
2         2.064330   -15.424000  -423.212000    -43.99948  ... -0.901268   
3         1.228467    -8.733602  -413.451300    -53.88054  ... -0.897777   
4         0.319165    10.939650  -389.706500    -63.49560  ... -0.892533   
...            ...          ...          ...          ...  ...       ...   
13165     9.877092   406.877300    83.288230   -233.31810  ... -0.262605   
13166   198.896500    73.391430     9.156261    299.95340  ... -0.114448   
13167    -8.981656   -60.419030  -147.604300   -261.47640  ...  0.492375   
13168    -1.901407   635.999900    66.554560     67.76263  ... -0.124552   
13169    63.869320   178.117200  -200.984500     41.27629  ...  0.100648   

        Wrist Pos X   Wrist Pos Y   Wrist Pos Z   Elbow pos X   Elbow Pos Y  \
0         74.461410      202.4562      44.28679      282.6278     140.97120   
1         73.600600      197.4387      43.54082      280.6384     142.54010   
2         73.195370      191.7202      42.73155      279.7357     148.13820   
3         72.686710      186.1613      41.98006      278.6890     152.52150   
4         72.376450      180.5758      41.30936      277.8844     155.16360   
...             ...           ...           ...           ...           ...   
13165     -6.809187      218.9017      62.84002      103.1046      69.87738   
13166     41.626430      104.6813     240.26050      173.8254     -93.86032   
13167           NaN           NaN           NaN           NaN           NaN   
13168     84.374400      267.8775      73.51348      192.1557     194.25960   
13169     50.344800      199.7355     125.65570      170.9249      44.57318   

        Elbow Pos Z   Grab Strenth   Grab Angle   Pinch Strength  
0          203.6773       0.000000     0.626321         0.000000  
1          206.7526       0.000000     0.636596         0.000000  
2          209.9330       0.000000     0.701644         0.000000  
3          212.1112       0.000000     0.786548         0.000000  
4          213.4511       0.000000     0.816478         0.000000  
...             ...            ...          ...              ...  
13165      261.9435       0.000000     0.282397         0.000000  
13166      375.0492       0.000000     0.275448         0.000000  
13167           NaN       0.000000     0.321349         0.000000  
13168      312.0404       0.205253     1.355023         0.380650  
13169      313.8539       0.000000     2.637807         0.660062  

[13170 rows x 22 columns]

In [15]:
#We need to achive 31199 rows, therefore I will do it again. 

dfAddHandRight = HandRight.sample(n=4390,replace=True)

#and merge it with the dfHandRight1 df that already exist

framesHandToAdd=[dfHandRight2,dfAddHandRight]

dfHandRight3 = pd.concat(framesHandToAdd)
dfHandRight3=dfHandRight3.reset_index()
dfHandRight3.drop(['index'],axis='columns', inplace=True)
dfHandRight3

Time   Frame ID  Hand Type   # hands   Position X   Position Y  \
0      135.3194      16289      right         1     11.19190     195.3075   
1      135.3357      16291      right         1     10.53308     189.0297   
2      135.3524      16293      right         1     10.24300     181.9816   
3      135.3690      16295      right         1     10.02061     174.8592   
4      135.3857      16297      right         1     10.12027     167.8396   
...         ...        ...        ...       ...          ...          ...   
17555  206.5291      24495      right         1    -19.26550     109.9357   
17556  161.6937      19330      right         1     54.53690     124.8134   
17557  144.9656      17401      right         1     54.56993     167.6867   
17558  136.4354      16418      right         1     -7.00864     188.5157   
17559  197.8142      23493      right         1    -92.45758     243.6802   

        Position Z   Velocity X   Velocity Y   Velocity Z  ...       Yaw  \
0         3.661716   -43.587440   -309.83740    -77.18515  ... -0.904311   
1         2.829388   -30.403600   -363.00330    -43.79815  ... -0.901862   
2         2.064330   -15.424000   -423.21200    -43.99948  ... -0.901268   
3         1.228467    -8.733602   -413.45130    -53.88054  ... -0.897777   
4         0.319165    10.939650   -389.70650    -63.49560  ... -0.892533   
...            ...          ...          ...          ...  ...       ...   
17555     5.957397    90.533600    236.73790   -143.20760  ... -0.073448   
17556    49.878300    -6.889000    176.01930    -76.65077  ... -0.000403   
17557    48.587520  -131.899400     80.35982    -36.74503  ...  0.047148   
17558   -12.169110    12.161270   -271.35090    151.66100  ... -0.788583   
17559   136.303900  -226.560000    -66.01006     80.49596  ... -0.309549   

        Wrist Pos X   Wrist Pos Y   Wrist Pos Z   Elbow pos X   Elbow Pos Y  \
0          74.46141      202.4562      44.28679      282.6278     140.97120   
1          73.60060      197.4387      43.54082      280.6384     142.54010   
2          73.19537      191.7202      42.73155      279.7357     148.13820   
3          72.68671      186.1613      41.98006      278.6890     152.52150   
4          72.37645      180.5758      41.30936      277.8844     155.16360   
...             ...           ...           ...           ...           ...   
17555           NaN           NaN           NaN           NaN           NaN   
17556      67.58176      125.1643     126.24730      218.4123     105.36890   
17557      64.19629      135.8635     118.08780      177.9696      35.31454   
17558      51.85654      192.0767      35.00497      240.9175     108.75280   
17559     -59.41632      212.4266     197.80020      136.8094      49.91672   

        Elbow Pos Z   Grab Strenth   Grab Angle   Pinch Strength  
0          203.6773            0.0     0.626321              0.0  
1          206.7526            0.0     0.636596              0.0  
2          209.9330            0.0     0.701644              0.0  
3          212.1112            0.0     0.786548              0.0  
4          213.4511            0.0     0.816478              0.0  
...             ...            ...          ...              ...  
17555           NaN            0.0     0.888781              0.0  
17556      355.6736            0.0     0.180565              0.0  
17557      344.8638            0.0     0.197277              0.0  
17558      207.8943            0.0     0.184295              0.0  
17559      297.6526            0.0     0.139772              0.0  

[17560 rows x 22 columns]

In [16]:
#We need to achive 31199 rows, therefore I will do it again. 

dfAddHandRight = HandRight.sample(n=4390,replace=True)

#and merge it with the dfHandRight1 df that already exist

framesHandToAdd=[dfHandRight3,dfAddHandRight]

dfHandRight4 = pd.concat(framesHandToAdd)
dfHandRight4=dfHandRight4.reset_index()
dfHandRight4.drop(['index'],axis='columns', inplace=True)
dfHandRight4

Time   Frame ID  Hand Type   # hands   Position X   Position Y  \
0      135.3194      16289      right         1    11.191900     195.3075   
1      135.3357      16291      right         1    10.533080     189.0297   
2      135.3524      16293      right         1    10.243000     181.9816   
3      135.3690      16295      right         1    10.020610     174.8592   
4      135.3857      16297      right         1    10.120270     167.8396   
...         ...        ...        ...       ...          ...          ...   
21945  202.5132      24033      right         1    -3.811424     144.6778   
21946  184.3857      21946      right         1    -1.377341     206.5292   
21947  159.7773      19109      right         1    55.505570     168.3702   
21948  195.7486      23255      right         1   -14.219710     191.7849   
21949  188.1676      22381      right         1     1.837306     243.6463   

        Position Z   Velocity X   Velocity Y   Velocity Z  ...       Yaw  \
0         3.661716   -43.587440   -309.83740    -77.18515  ... -0.904311   
1         2.829388   -30.403600   -363.00330    -43.79815  ... -0.901862   
2         2.064330   -15.424000   -423.21200    -43.99948  ... -0.901268   
3         1.228467    -8.733602   -413.45130    -53.88054  ... -0.897777   
4         0.319165    10.939650   -389.70650    -63.49560  ... -0.892533   
...            ...          ...          ...          ...  ...       ...   
21945    79.575310    31.775630     92.24528   -520.16430  ... -0.012283   
21946   -27.368110    34.263850    194.86230     83.57300  ... -0.340112   
21947    46.459720  -788.680700    -74.89828    -64.08276  ... -0.009646   
21948    60.760110  -100.278100    371.43020    665.91330  ... -0.146587   
21949    15.860590    17.124160    115.41520    522.59630  ... -0.270429   

        Wrist Pos X   Wrist Pos Y   Wrist Pos Z   Elbow pos X   Elbow Pos Y  \
0         74.461410      202.4562      44.28679      282.6278    140.971200   
1         73.600600      197.4387      43.54082      280.6384    142.540100   
2         73.195370      191.7202      42.73155      279.7357    148.138200   
3         72.686710      186.1613      41.98006      278.6890    152.521500   
4         72.376450      180.5758      41.30936      277.8844    155.163600   
...             ...           ...           ...           ...           ...   
21945           NaN           NaN           NaN           NaN           NaN   
21946     25.285280      156.7189      24.60232      113.5616    -28.060980   
21947     68.265100      156.1161     121.89340      181.5443     73.503910   
21948      8.196917      157.0745     125.07100      171.8655     15.269780   
21949     21.062510      176.4078      47.31981      124.1824     -6.142286   

        Elbow Pos Z   Grab Strenth   Grab Angle   Pinch Strength  
0          203.6773       0.000000     0.626321         0.000000  
1          206.7526       0.000000     0.636596         0.000000  
2          209.9330       0.000000     0.701644         0.000000  
3          212.1112       0.000000     0.786548         0.000000  
4          213.4511       0.000000     0.816478         0.000000  
...             ...            ...          ...              ...  
21945           NaN       0.000000     0.688216         0.000000  
21946      203.6115       0.067548     1.237484         0.118163  
21947      358.3590       0.000000     0.254823         0.000000  
21948      289.0072       0.000000     0.017119         0.000000  
21949      220.5895       0.000000     0.537352         0.000000  

[21950 rows x 22 columns]

In [17]:
#We need to achive 31199 rows, therefore I will do it again. 

dfAddHandRight = HandRight.sample(n=4390,replace=True)

#and merge it with the dfHandRight1 df that already exist

framesHandToAdd=[dfHandRight4,dfAddHandRight]

dfHandRight5 = pd.concat(framesHandToAdd)
dfHandRight5=dfHandRight5.reset_index()
dfHandRight5.drop(['index'],axis='columns', inplace=True)
dfHandRight5

Time   Frame ID  Hand Type   # hands   Position X   Position Y  \
0      135.3194      16289      right         1    11.191900     195.3075   
1      135.3357      16291      right         1    10.533080     189.0297   
2      135.3524      16293      right         1    10.243000     181.9816   
3      135.3690      16295      right         1    10.020610     174.8592   
4      135.3857      16297      right         1    10.120270     167.8396   
...         ...        ...        ...       ...          ...          ...   
26335  191.4342      22757      right         1    61.732360     188.8982   
26336  159.8273      19114      right         1    26.810590     162.0648   
26337  183.2694      21818      right         1    -4.070035     199.8774   
26338  152.0473      18217      right         2    46.948720     219.2077   
26339  177.6712      21172      right         1    33.957280     120.6595   

        Position Z   Velocity X   Velocity Y   Velocity Z  ...       Yaw  \
0         3.661716   -43.587440   -309.83740   -77.185150  ... -0.904311   
1         2.829388   -30.403600   -363.00330   -43.798150  ... -0.901862   
2         2.064330   -15.424000   -423.21200   -43.999480  ... -0.901268   
3         1.228467    -8.733602   -413.45130   -53.880540  ... -0.897777   
4         0.319165    10.939650   -389.70650   -63.495600  ... -0.892533   
...            ...          ...          ...          ...  ...       ...   
26335    51.171810  -747.983900   -314.41070   -74.166660  ... -0.175354   
26336    44.273390  -545.173900   -124.24010   -51.831430  ... -0.067791   
26337   -28.120940    69.583570   -166.18230  -172.094600  ... -0.358456   
26338    52.475650    11.503880   -118.51060    -5.338694  ... -0.006189   
26339    13.759410   -11.321070     78.33913    -4.770303  ...  0.321371   

        Wrist Pos X   Wrist Pos Y   Wrist Pos Z   Elbow pos X   Elbow Pos Y  \
0         74.461410      202.4562      44.28679     282.62780     140.97120   
1         73.600600      197.4387      43.54082     280.63840     142.54010   
2         73.195370      191.7202      42.73155     279.73570     148.13820   
3         72.686710      186.1613      41.98006     278.68900     152.52150   
4         72.376450      180.5758      41.30936     277.88440     155.16360   
...             ...           ...           ...           ...           ...   
26335     82.777830      232.6284     108.26990     244.93180     204.49510   
26336     43.398090      153.2499     119.36740     163.83830      69.63158   
26337     24.516550      149.4327      22.29400     135.14470     -17.48877   
26338     61.480490      213.4467     128.13200     197.47280     122.11530   
26339      9.237543      110.0591      85.39525      19.97346      38.39731   

        Elbow Pos Z   Grab Strenth   Grab Angle   Pinch Strength  
0          203.6773       0.000000     0.626321              0.0  
1          206.7526       0.000000     0.636596              0.0  
2          209.9330       0.000000     0.701644              0.0  
3          212.1112       0.000000     0.786548              0.0  
4          213.4511       0.000000     0.816478              0.0  
...             ...            ...          ...              ...  
26335      324.7095       0.210283     1.980744              0.0  
26336      351.8928       0.000000     0.283186              0.0  
26337      206.3572       0.000000     0.000000              0.0  
26338      348.7052       0.000000     0.613255              0.0  
26339      348.2132       0.000000     1.018620              0.0  

[26340 rows x 22 columns]

In [18]:
#We need to achive 31199 rows, therefore I will do it again. 

dfAddHandRight = HandRight.sample(n=4390,replace=True)

#and merge it with the dfHandRight1 df that already exist

framesHandToAdd=[dfHandRight5,dfAddHandRight]

dfHandRight6 = pd.concat(framesHandToAdd)
dfHandRight6=dfHandRight6.reset_index()
dfHandRight6.drop(['index'],axis='columns', inplace=True)
dfHandRight6

Time   Frame ID  Hand Type   # hands   Position X   Position Y  \
0      135.3194      16289      right         1    11.191900     195.3075   
1      135.3357      16291      right         1    10.533080     189.0297   
2      135.3524      16293      right         1    10.243000     181.9816   
3      135.3690      16295      right         1    10.020610     174.8592   
4      135.3857      16297      right         1    10.120270     167.8396   
...         ...        ...        ...       ...          ...          ...   
30725  198.1643      23533      right         1  -139.282600     157.7006   
30726  168.9908      20171      right         1     5.891465     195.2361   
30727  172.3232      20555      right         1    -0.713537     125.7674   
30728  189.5339      22538      right         1   -20.136130     184.9273   
30729  181.6699      21633      right         1    17.667790     254.1207   

        Position Z   Velocity X   Velocity Y   Velocity Z  ...       Yaw  \
0         3.661716   -43.587440    -309.8374    -77.18515  ... -0.904311   
1         2.829388   -30.403600    -363.0033    -43.79815  ... -0.901862   
2         2.064330   -15.424000    -423.2120    -43.99948  ... -0.901268   
3         1.228467    -8.733602    -413.4513    -53.88054  ... -0.897777   
4         0.319165    10.939650    -389.7065    -63.49560  ... -0.892533   
...            ...          ...          ...          ...  ...       ...   
30725    36.502460   548.625400     114.5270   -410.23840  ... -0.316946   
30726   -27.223560    17.201130    -442.4770   -111.08230  ... -0.131331   
30727   125.440100   -74.995000     544.1633    388.47730  ... -0.113038   
30728    16.755970 -1181.068000    -423.4577     16.79097  ... -0.203634   
30729    17.989770   -70.760310     248.7132     45.90860  ... -0.307307   

        Wrist Pos X   Wrist Pos Y   Wrist Pos Z   Elbow pos X   Elbow Pos Y  \
0         74.461410     202.45620      44.28679     282.62780     140.97120   
1         73.600600     197.43870      43.54082     280.63840     142.54010   
2         73.195370     191.72020      42.73155     279.73570     148.13820   
3         72.686710     186.16130      41.98006     278.68900     152.52150   
4         72.376450     180.57580      41.30936     277.88440     155.16360   
...             ...           ...           ...           ...           ...   
30725   -105.892500     146.08000     105.07000      70.61903      78.68495   
30726     21.137260     131.70630      12.83376      95.93491       3.11302   
30727     19.040770      92.87824     192.01490     158.22210     -49.14401   
30728      5.625952     157.17380      83.17406     152.53170      12.17599   
30729     44.984870     204.81080      70.15736     156.14630      18.53131   

        Elbow Pos Z   Grab Strenth   Grab Angle   Pinch Strength  
0          203.6773       0.000000     0.626321              0.0  
1          206.7526       0.000000     0.636596              0.0  
2          209.9330       0.000000     0.701644              0.0  
3          212.1112       0.000000     0.786548              0.0  
4          213.4511       0.000000     0.816478              0.0  
...             ...            ...          ...              ...  
30725      303.0924       0.000000     0.020552              0.0  
30726      241.6023       0.000000     0.366927              0.0  
30727      378.5980       0.000000     0.510735              0.0  
30728      259.7316       0.000000     0.871383              0.0  
30729      234.2419       0.114024     1.732610              1.0  

[30730 rows x 22 columns]

In [19]:
print("Now I need to add more",len(dfAlone.index)-len(dfHandRight6.index),"rows. Let's do it:")

Now I need to add more 469 rows. Let's do it:


In [20]:
dfAddHandRight = HandRight.sample(n=len(dfAlone.index)-len(dfHandRight6.index),replace=True)

#and merge it with the dfHandRight1 df that already exist

framesHandToAdd=[dfHandRight6,dfAddHandRight]

dfHandRight7 = pd.concat(framesHandToAdd)
dfHandRight7=dfHandRight7.reset_index()
dfHandRight7.drop(['index'],axis='columns', inplace=True)
dfHandRight7

Time   Frame ID  Hand Type   # hands   Position X   Position Y  \
0      135.3194      16289      right         1    11.191900     195.3075   
1      135.3357      16291      right         1    10.533080     189.0297   
2      135.3524      16293      right         1    10.243000     181.9816   
3      135.3690      16295      right         1    10.020610     174.8592   
4      135.3857      16297      right         1    10.120270     167.8396   
...         ...        ...        ...       ...          ...          ...   
31194  146.3819      17564      right         1    33.473550     251.3325   
31195  174.7721      20838      right         1    -6.631704     196.9134   
31196  158.1116      18916      right         1    34.563900     148.2518   
31197  142.1167      17073      right         1    29.652120     211.0022   
31198  178.8880      21312      right         1    42.287410     112.4894   

        Position Z   Velocity X   Velocity Y   Velocity Z  ...       Yaw  \
0         3.661716   -43.587440   -309.83740   -77.185150  ... -0.904311   
1         2.829388   -30.403600   -363.00330   -43.798150  ... -0.901862   
2         2.064330   -15.424000   -423.21200   -43.999480  ... -0.901268   
3         1.228467    -8.733602   -413.45130   -53.880540  ... -0.897777   
4         0.319165    10.939650   -389.70650   -63.495600  ... -0.892533   
...            ...          ...          ...          ...  ...       ...   
31194    40.301140   195.882500   -166.96160   -15.776070  ... -0.075368   
31195   100.530100   -52.677230   -179.41520   161.463300  ... -0.003909   
31196    12.535700   -68.443780    117.20920  -133.305900  ... -0.068710   
31197    51.340070   211.696900    220.66290    -1.609264  ... -0.081846   
31198     6.011418   -43.370570    -62.03915    -0.785919  ...  0.313903   

        Wrist Pos X   Wrist Pos Y   Wrist Pos Z   Elbow pos X   Elbow Pos Y  \
0         74.461410      202.4562      44.28679     282.62780    140.971200   
1         73.600600      197.4387      43.54082     280.63840    142.540100   
2         73.195370      191.7202      42.73155     279.73570    148.138200   
3         72.686710      186.1613      41.98006     278.68900    152.521500   
4         72.376450      180.5758      41.30936     277.88440    155.163600   
...             ...           ...           ...           ...           ...   
31194     50.222030      221.9234     109.75110     170.70470     74.525470   
31195      9.116628      164.2832     167.49430     186.08370     -1.664056   
31196     51.422520      151.3784      87.68416     164.08020     75.865440   
31197     45.855510      169.8300     114.10710     177.01130     41.816100   
31198     20.603490       72.6887      66.69569      35.22388     31.051900   

        Elbow Pos Z   Grab Strenth   Grab Angle   Pinch Strength  
0          203.6773       0.000000     0.626321         0.000000  
1          206.7526       0.000000     0.636596         0.000000  
2          209.9330       0.000000     0.701644         0.000000  
3          212.1112       0.000000     0.786548         0.000000  
4          213.4511       0.000000     0.816478         0.000000  
...             ...            ...          ...              ...  
31194      307.3716       0.000000     0.222160         0.000000  
31195      292.1238       0.000000     1.110166         0.000000  
31196      326.1358       0.000000     0.438989         0.000000  
31197      315.2142       0.000000     2.651099         0.624211  
31198      335.7246       0.706638     2.671068         0.000000  

[31199 rows x 22 columns]

In [21]:
#Let's check the size of the new df of HandRight and dfAlone:
print(len(dfAlone.index))
HandRight=dfHandRight7
print(len(HandRight.index)) 

31199
31199


## Merge both dataframes

It is possible to see that the lenght of both df is equal , therefore I will merge them so that after each row from dfAlone,<br>
There will be a row from HandRight df.

In [22]:
#Create new empty df with the names of the columns

dfAloneAndRight=pd.DataFrame(columns=('Time', ' Frame ID',' Hand Type',' # hands', ' Position X', ' Position Y', ' Position Z', ' Velocity X',
             ' Velocity Y', ' Velocity Z',' Pitch',' Roll',' Yaw', ' Wrist Pos X', ' Wrist Pos Y', ' Wrist Pos Z', ' Elbow pos X',
             ' Elbow Pos Y',' Elbow Pos Z',' Grab Strenth', ' Grab Angle', ' Pinch Strength','HandPosition_Label'))
dfAloneAndRight

Empty DataFrame
Columns: [Time,  Frame ID,  Hand Type,  # hands,  Position X,  Position Y,  Position Z,  Velocity X,  Velocity Y,  Velocity Z,  Pitch,  Roll,  Yaw,  Wrist Pos X,  Wrist Pos Y,  Wrist Pos Z,  Elbow pos X,  Elbow Pos Y,  Elbow Pos Z,  Grab Strenth,  Grab Angle,  Pinch Strength, HandPosition_Label]
Index: []

[0 rows x 23 columns]

In [23]:
dfAloneAndRight=dfAlone.loc[[0]]
i=0
while i<31199:
    dfAloneAndRight=dfAloneAndRight.append(dfAlone.loc[[i]])
    dfAloneAndRight=dfAloneAndRight.append(HandRight.loc[[i]])
    i=i+1
dfAloneAndRight

Time   Frame ID  Hand Type   # hands   Position X   Position Y  \
0      532.8434     151238       left         1    -30.37894     225.4433   
0      532.8434     151238       left         1    -30.37894     225.4433   
0      135.3194      16289      right         1     11.19190     195.3075   
1      532.8600     151240       left         1    -32.18705     229.7671   
1      135.3357      16291      right         1     10.53308     189.0297   
...         ...        ...        ...       ...          ...          ...   
31196  158.1116      18916      right         1     34.56390     148.2518   
31197  764.5587      59989       left         1   -125.13520     215.6353   
31197  142.1167      17073      right         1     29.65212     211.0022   
31198  764.5913      59991       left         1   -123.99810     215.1558   
31198  178.8880      21312      right         1     42.28741     112.4894   

        Position Z   Velocity X   Velocity Y   Velocity Z  ...       Yaw  \
0        46.163120    -83.61432   271.978200    26.200460  ...  1.558521   
0        46.163120    -83.61432   271.978200    26.200460  ...  1.558521   
0         3.661716    -43.58744  -309.837400   -77.185150  ... -0.904311   
1        46.527040   -118.26000   230.771500    11.701220  ...  1.539598   
1         2.829388    -30.40360  -363.003300   -43.798150  ... -0.901862   
...            ...          ...          ...          ...  ...       ...   
31196    12.535700    -68.44378   117.209200  -133.305900  ... -0.068710   
31197    23.802480     25.56741    -5.648659    21.284490  ...  0.368866   
31197    51.340070    211.69690   220.662900    -1.609264  ... -0.081846   
31198    24.269450     30.36909   -15.233250     6.734551  ...  0.373039   
31198     6.011418    -43.37057   -62.039150    -0.785919  ...  0.313903   

        Wrist Pos X   Wrist Pos Y   Wrist Pos Z   Elbow pos X   Elbow Pos Y  \
0         -87.51802      183.6657      37.15565    -301.85890      54.37961   
0         -87.51802      183.6657      37.15565    -301.85890      54.37961   
0          74.46141      202.4562      44.28679     282.62780     140.97120   
1         -87.13968      184.9756      38.34803    -293.86520      44.03124   
1          73.60060      197.4387      43.54082     280.63840     142.54010   
...             ...           ...           ...           ...           ...   
31196      51.42252      151.3784      87.68416     164.08020      75.86544   
31197    -143.84820      148.1024      46.72861    -248.27680     -84.26936   
31197      45.85551      169.8300     114.10710     177.01130      41.81610   
31198    -142.68390      147.4703      46.75579    -247.41210     -84.85361   
31198      20.60349       72.6887      66.69569      35.22388      31.05190   

        Elbow Pos Z   Grab Strenth   Grab Angle   Pinch Strength  
0          2.226311       0.000000     0.032902         0.000000  
0          2.226311       0.000000     0.032902         0.000000  
0        203.677300       0.000000     0.626321         0.000000  
1          2.624075       0.000000     0.136581         0.000000  
1        206.752600       0.000000     0.636596         0.000000  
...             ...            ...          ...              ...  
31196    326.135800       0.000000     0.438989         0.000000  
31197    103.468300       0.097735     1.988376         0.000000  
31197    315.214200       0.000000     2.651099         0.624211  
31198    103.138000       0.065015     1.989704         0.016533  
31198    335.724600       0.706638     2.671068         0.000000  

[62399 rows x 22 columns]

### Define label alone as number 2

In [24]:
#I will define the whole column 'HandPosition_Label' as 2 (Alone position)         
dfAloneAndRight['HandPosition_Label']=2
dfAloneAndRight

Time   Frame ID  Hand Type   # hands   Position X   Position Y  \
0      532.8434     151238       left         1    -30.37894     225.4433   
0      532.8434     151238       left         1    -30.37894     225.4433   
0      135.3194      16289      right         1     11.19190     195.3075   
1      532.8600     151240       left         1    -32.18705     229.7671   
1      135.3357      16291      right         1     10.53308     189.0297   
...         ...        ...        ...       ...          ...          ...   
31196  158.1116      18916      right         1     34.56390     148.2518   
31197  764.5587      59989       left         1   -125.13520     215.6353   
31197  142.1167      17073      right         1     29.65212     211.0022   
31198  764.5913      59991       left         1   -123.99810     215.1558   
31198  178.8880      21312      right         1     42.28741     112.4894   

        Position Z   Velocity X   Velocity Y   Velocity Z  ...   Wrist Pos X  \
0        46.163120    -83.61432   271.978200    26.200460  ...     -87.51802   
0        46.163120    -83.61432   271.978200    26.200460  ...     -87.51802   
0         3.661716    -43.58744  -309.837400   -77.185150  ...      74.46141   
1        46.527040   -118.26000   230.771500    11.701220  ...     -87.13968   
1         2.829388    -30.40360  -363.003300   -43.798150  ...      73.60060   
...            ...          ...          ...          ...  ...           ...   
31196    12.535700    -68.44378   117.209200  -133.305900  ...      51.42252   
31197    23.802480     25.56741    -5.648659    21.284490  ...    -143.84820   
31197    51.340070    211.69690   220.662900    -1.609264  ...      45.85551   
31198    24.269450     30.36909   -15.233250     6.734551  ...    -142.68390   
31198     6.011418    -43.37057   -62.039150    -0.785919  ...      20.60349   

        Wrist Pos Y   Wrist Pos Z   Elbow pos X   Elbow Pos Y   Elbow Pos Z  \
0          183.6657      37.15565    -301.85890      54.37961      2.226311   
0          183.6657      37.15565    -301.85890      54.37961      2.226311   
0          202.4562      44.28679     282.62780     140.97120    203.677300   
1          184.9756      38.34803    -293.86520      44.03124      2.624075   
1          197.4387      43.54082     280.63840     142.54010    206.752600   
...             ...           ...           ...           ...           ...   
31196      151.3784      87.68416     164.08020      75.86544    326.135800   
31197      148.1024      46.72861    -248.27680     -84.26936    103.468300   
31197      169.8300     114.10710     177.01130      41.81610    315.214200   
31198      147.4703      46.75579    -247.41210     -84.85361    103.138000   
31198       72.6887      66.69569      35.22388      31.05190    335.724600   

        Grab Strenth   Grab Angle   Pinch Strength  HandPosition_Label  
0           0.000000     0.032902         0.000000                   2  
0           0.000000     0.032902         0.000000                   2  
0           0.000000     0.626321         0.000000                   2  
1           0.000000     0.136581         0.000000                   2  
1           0.000000     0.636596         0.000000                   2  
...              ...          ...              ...                 ...  
31196       0.000000     0.438989         0.000000                   2  
31197       0.097735     1.988376         0.000000                   2  
31197       0.000000     2.651099         0.624211                   2  
31198       0.065015     1.989704         0.016533                   2  
31198       0.706638     2.671068         0.000000                   2  

[62399 rows x 23 columns]

In [26]:
#instead of run this cell every time, I will save it as csv file
dfAloneAndRight.to_csv(r'C:\Users\yuval\Datasets\HandsPositions\dfAloneAndRight.csv')


In [27]:
#import again the csv file
dfAloneAndRight = pd.read_csv(r"C:\Users\yuval\Datasets\HandsPositions\dfAloneAndRight.csv")
dfAloneAndRight

Unnamed: 0      Time   Frame ID  Hand Type   # hands   Position X  \
0               0  532.8434     151238       left         1    -30.37894   
1               0  532.8434     151238       left         1    -30.37894   
2               0  135.3194      16289      right         1     11.19190   
3               1  532.8600     151240       left         1    -32.18705   
4               1  135.3357      16291      right         1     10.53308   
...           ...       ...        ...        ...       ...          ...   
62394       31196  158.1116      18916      right         1     34.56390   
62395       31197  764.5587      59989       left         1   -125.13520   
62396       31197  142.1167      17073      right         1     29.65212   
62397       31198  764.5913      59991       left         1   -123.99810   
62398       31198  178.8880      21312      right         1     42.28741   

        Position Y   Position Z   Velocity X   Velocity Y  ...   Wrist Pos X  \
0         225.4433    46.163120    -83.61432   271.978200  ...     -87.51802   
1         225.4433    46.163120    -83.61432   271.978200  ...     -87.51802   
2         195.3075     3.661716    -43.58744  -309.837400  ...      74.46141   
3         229.7671    46.527040   -118.26000   230.771500  ...     -87.13968   
4         189.0297     2.829388    -30.40360  -363.003300  ...      73.60060   
...            ...          ...          ...          ...  ...           ...   
62394     148.2518    12.535700    -68.44378   117.209200  ...      51.42252   
62395     215.6353    23.802480     25.56741    -5.648659  ...    -143.84820   
62396     211.0022    51.340070    211.69690   220.662900  ...      45.85551   
62397     215.1558    24.269450     30.36909   -15.233250  ...    -142.68390   
62398     112.4894     6.011418    -43.37057   -62.039150  ...      20.60349   

        Wrist Pos Y   Wrist Pos Z   Elbow pos X   Elbow Pos Y   Elbow Pos Z  \
0          183.6657      37.15565    -301.85890      54.37961      2.226311   
1          183.6657      37.15565    -301.85890      54.37961      2.226311   
2          202.4562      44.28679     282.62780     140.97120    203.677300   
3          184.9756      38.34803    -293.86520      44.03124      2.624075   
4          197.4387      43.54082     280.63840     142.54010    206.752600   
...             ...           ...           ...           ...           ...   
62394      151.3784      87.68416     164.08020      75.86544    326.135800   
62395      148.1024      46.72861    -248.27680     -84.26936    103.468300   
62396      169.8300     114.10710     177.01130      41.81610    315.214200   
62397      147.4703      46.75579    -247.41210     -84.85361    103.138000   
62398       72.6887      66.69569      35.22388      31.05190    335.724600   

        Grab Strenth   Grab Angle   Pinch Strength  HandPosition_Label  
0           0.000000     0.032902         0.000000                   2  
1           0.000000     0.032902         0.000000                   2  
2           0.000000     0.626321         0.000000                   2  
3           0.000000     0.136581         0.000000                   2  
4           0.000000     0.636596         0.000000                   2  
...              ...          ...              ...                 ...  
62394       0.000000     0.438989         0.000000                   2  
62395       0.097735     1.988376         0.000000                   2  
62396       0.000000     2.651099         0.624211                   2  
62397       0.065015     1.989704         0.016533                   2  
62398       0.706638     2.671068         0.000000                   2  

[62399 rows x 24 columns]

In [28]:
dfAloneAndRight=dfAloneAndRight.drop(['Unnamed: 0'], axis=1)

## Merge dfAloneAndRight and dfSync_Spontan
Now After I have 2 dataframe dfAloneAndRight and dfSync_Spontan- let's merge them both into one general df

In [29]:
#Merge all Alone dataframes into one, with concat method as we learned previous semester

frames_LeftAndRight=[dfAloneAndRight,dfSync_Spontan]
dfGeneral = pd.concat(frames_LeftAndRight)
dfGeneral=dfGeneral.reset_index()
dfGeneral.drop(['index'],axis='columns', inplace=True)
dfGeneral

Time   Frame ID  Hand Type   # hands   Position X   Position Y  \
0       532.8434     151238       left         1    -30.37894     225.4433   
1       532.8434     151238       left         1    -30.37894     225.4433   
2       135.3194      16289      right         1     11.19190     195.3075   
3       532.8600     151240       left         1    -32.18705     229.7671   
4       135.3357      16291      right         1     10.53308     189.0297   
...          ...        ...        ...       ...          ...          ...   
199288  856.2631      67309      right         2    101.91230     212.0264   
199289  856.2958      67311       left         2    -81.57359     183.5054   
199290  856.2958      67311      right         2    103.84750     214.7484   
199291  856.3295      67314       left         2    -84.05124     187.8450   
199292  856.3295      67314      right         2    106.11880     218.0190   

         Position Z   Velocity X   Velocity Y   Velocity Z  ...   Wrist Pos X  \
0         46.163120    -83.61432    271.97820     26.20046  ...     -87.51802   
1         46.163120    -83.61432    271.97820     26.20046  ...     -87.51802   
2          3.661716    -43.58744   -309.83740    -77.18515  ...      74.46141   
3         46.527040   -118.26000    230.77150     11.70122  ...     -87.13968   
4          2.829388    -30.40360   -363.00330    -43.79815  ...      73.60060   
...             ...          ...          ...          ...  ...           ...   
199288    42.535680     79.95810    113.41030    190.77140  ...     131.85900   
199289    -7.185198    -63.24731    121.75270     62.95970  ...    -109.72500   
199290    47.070820     61.11773     95.13531    155.74530  ...     134.04160   
199291    -5.172269    -57.17249    109.20360     58.48146  ...    -109.51670   
199292    53.588680     25.46704     72.26683    190.16360  ...     137.07080   

         Wrist Pos Y   Wrist Pos Z   Elbow pos X   Elbow Pos Y   Elbow Pos Z  \
0           183.6657      37.15565     -301.8589      54.37961      2.226311   
1           183.6657      37.15565     -301.8589      54.37961      2.226311   
2           202.4562      44.28679      282.6278     140.97120    203.677300   
3           184.9756      38.34803     -293.8652      44.03124      2.624075   
4           197.4387      43.54082      280.6384     142.54010    206.752600   
...              ...           ...           ...           ...           ...   
199288      149.6264      66.09853      283.6185     -58.03491     95.298680   
199289      130.7078      34.03360     -295.5957     -14.97817    140.026300   
199290      151.5924      68.17693      288.9688     -54.13820     94.239590   
199291      131.6489      33.32945     -282.9649     -31.35462    135.076100   
199292      154.2158      71.46925      289.6567     -53.67185     93.987310   

         Grab Strenth   Grab Angle   Pinch Strength  HandPosition_Label  
0                 0.0     0.032902              0.0                   2  
1                 0.0     0.032902              0.0                   2  
2                 0.0     0.626321              0.0                   2  
3                 0.0     0.136581              0.0                   2  
4                 0.0     0.636596              0.0                   2  
...               ...          ...              ...                 ...  
199288            0.0     0.000000              0.0                   0  
199289            0.0     0.000000              0.0                   0  
199290            0.0     0.000000              0.0                   0  
199291            0.0     0.000000              0.0                   0  
199292            0.0     0.000000              0.0                   0  

[199293 rows x 23 columns]

## Consolidation of two records left&right 
As it already explained in class, each 2 rows of left&right hand is actually one frame, therefore I will turn each 2 rows into one.

In [ ]:
#function to combine 2 consecutive rows

liR = ["{}{}".format(ind,'Right') for ind in dfGeneral.columns] #appending "right" to the end of each column name
liL = ["{}{}".format(ind,'Left') for ind in dfGeneral.columns]
LiTotal = liR + liL

#for j in range(len(totalList)): #Iterating over every dataframe in the list
flattenedDf = pd.DataFrame(columns = LiTotal) #empty df
  #  totalList[j] = totalList[j].reset_index() #resetting index of the totalList[j] for merging
   # totalList[j] = totalList[j].drop(['index'], axis = 1)
    
for i in range(0, len(dfGeneral), 2): #Iterating over every record in the dataframe with jumps of 2 (due to
    #every iteration using 2 consecutive rows for merging)
    if (dfGeneral.loc[i][' Hand Type'] == 'right'):
        right = i
        left = i+1
    else:
        left = i
        right = i+1
    
    rightList = dfGeneral.loc[right]
    rightList.index = liR
        
    leftList = dfGeneral.loc[left]
    leftList.index = liL
    
    completeRecord = pd.concat([rightList, leftList])
    flattenedDf = flattenedDf.append(completeRecord, ignore_index = True)
dfGeneral = flattenedDf

In [ ]:
#instead of run this cell every time, I will save it as csv file
#dfGeneral.to_csv(r'C:\Users\yuval\Datasets\HandsPositions\dfGeneral.csv')

In [ ]:
#import again the csv file
dfGeneral = pd.read_csv(r"C:\Users\yuval\Datasets\HandsPositions\dfGeneral.csv")
dfGeneral

# Data cleaning
## <b>know and organize the dataset</b>

In [ ]:
# check the shape of the DataFrame (rows, columns)
dfGeneral.shape 

**Explanation about the columns and rows:**<br>HHHHHHHHHHHHHHHHHHHHHHH

**What are the features?**<br>
60000 rows- every row represent list of pixels in the image.<br>
The first row of pixels in the image is the first 28 cells in a row in the df, the next 28 cells are the seconed row of pixels in the image, etc. .<br>
784 columns of pixels- After I set a uniform size for all the images, the size is 28*28, therefore there are 784 rows, plus 1 row .


**What is the response?**<br>
label-What item of clothing is this from the list I have written above (in range of 1-9).<br>

In [ ]:
#I want to check what the range of each column:
#by running both commands:
dfGeneral.max(axis = 0) 
dfGeneral.min(axis = 0) 
#Also can see it by use:
dfGeneral.describe()

**Conclusion**HHHHHHHHHHHH

In [ ]:
#We'll run the info() command to get a quick description of the data,particulary reguarding the total number of rows, 
#each attribute’s type, and the number of nonnull values.
dfGeneral.info()

## <b>Duplicate Values</b>

If there are duplicate values- means that specific image appear few times, therefore I would like to delete them.

In [ ]:
dfGeneral=dfGeneral.drop_duplicates()#By default, it removes duplicate rows based on all columns.

dfGeneral.shape

**Conclusion**

In [ ]:
#HHHHHHHHHHHHHHHHHHHHHHHHHThere were 60000 rows and now there are 59957 means we deleted 60000-59957=43 duplicate rows, we will write it at precentage:
print ("Percent of duplicate images I deleted:", (43*100)/60000,"%")

In [ ]:
# Histogram of all numeric features
dfGeneral.hist(bins=50, figsize=(20,15))
plt.show()

**Conclusion**HHHHHHHHHHHH

# Nan Values

In [ ]:
print(dfGeneral.isnull().sum())


In [ ]:
print(i)\

**Finish here**

# Columns I can drop+CorralationsHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHH+Unite by time- how many second will be together?

In [ ]:
#dfGeneral["beer_name"].value_counts() #to see how many kinds of beers there are in the df (56856)

#function to combine 2 consecutive rows
start_timeCombine = time.time()

liR = ["{}{}".format(ind,'Right') for ind in totalList[0].columns] #appending "right" to the end of each column name
liL = ["{}{}".format(ind,'Left') for ind in totalList[0].columns]
LiTotal = liR + liL

for j in range(len(totalList)): #Iterating over every dataframe in the list
    flattenedDf = pd.DataFrame(columns = LiTotal) #empty df
    totalList[j] = totalList[j].reset_index() #resetting index of the totalList[j] for merging
    totalList[j] = totalList[j].drop(['index'], axis = 1)
    
    for i in range(0, len(totalList[j]), 2): #Iterating over every record in the dataframe with jumps of 2 (due to
        #every iteration using 2 consecutive rows for merging)
        if (totalList[j].loc[i]['HandType'] == 'right'):
            right = i
            left = i+1
        else:
            left = i
            right = i+1
    
        rightList = totalList[j].loc[right]
        rightList.index = liR
        
        leftList = totalList[j].loc[left]
        leftList.index = liL
    
        completeRecord = pd.concat([rightList, leftList])
        flattenedDf = flattenedDf.append(completeRecord, ignore_index = True)
    totalList[j] = flattenedDf

 N = 3   # Say 3 more times

In [207]: pd.concat([df.add_suffix(1)] + 
                    [df[x+1:].reset_index(drop=True).add_suffix(x+2)
                     for x in range(N)], axis=1)

pd.concat([dfGeneral.add_suffix(1), df[1:].reset_index(drop=True).add_suffix(2)],
                    axis=1)

df1 = pd.concat([df.add_suffix('1'), df.shift(-1).add_suffix('2')], axis=1)
df1 = df.add_suffix('1').assign(**df.shift(-1).add_suffix('2'))

dfAlone.loc[[0]]

df = pd.DataFrame(columns=list('AB'))
df

df2 = pd.DataFrame([[5, 6], [7, 8]], columns=list('AB'), index=['x', 'y'])
df3=df2.loc[['x']]
df3


df.append(df3)
df

df = pd.DataFrame([[1, 2], [3, 4]], columns=list('AB'), index=['x', 'y'])
df2 = pd.DataFrame([[5, 6], [7, 8]], columns=list('AB'), index=['x', 'y'])
df=df.append(df2)
df

dfAloneAndRight.loc[[0]]=dfAlone.loc[[0]]

dfAloneAndRight

dfAloneAndRight.loc[[2]]

dfAlone.loc[[1]]

dfAloneAndRight.iloc[1]=dfAlone.iloc[1]
dfAloneAndRight.iloc[2]=dfAlone.iloc[1]
dfAloneAndRight

i=0
while i<5:
    dfAloneAndRight.loc[[2*i]]=dfAlone.loc[[i]]
    i += 1
    

dfAloneAndRight

i=0
#res = pd.DataFrame(columns=('lib', 'qty1', 'qty2'))
#dfAloneAndRight= pd.DataFrame(columns=('Time', ' Frame ID','Hand Type',' # hands', 'Position X', 'Position Y', 'Position Z', 'Velocity X',
#             'Velocity Y', 'Velocity Z','Pitch','Roll','Yaw', 'Wrist Pos X', 'Wrist Pos Y', 'Wrist Pos Z', 'Elbow pos X',
#             'Elbow Pos Y','Elbow Pos Z','Grab Strenth', 'Grab Angle', 'Pinch Strength'))
while i<5:#62398:

    #lineLeft=dfAlone.iloc[[i]]
    #lineRight=HandRight.iloc[[i]]
    #print(lineLeft)
    #print(lineRight)
    print(i)
    #dfAloneAndRight.iloc[[i]]=dfAlone.iloc[[i]]
    #dfAloneAndRight.iloc[[i]]=HandRight.iloc[[i]]
    #dfAloneAndRight.append(lineLeft.loc[i])
    j=(2*i)
    j=i
    f=((2*i)+1)
    print(j)
    print(f)
    
    
    dfAloneAndRight.iloc[[j]]=dfAlone.iloc[[i]]
    print("dfAlone.iloc[[i]]=")
    print(dfAlone.iloc[[i]])
    print("dfAloneAndRight.iloc[[j]]=")
    print(dfAloneAndRight.iloc[[j]])
    #dfAloneAndRight.iloc[[f]]=HandRight.iloc[[i]]
    #dfAloneAndRight.iloc[[j]]=dfAlone.iloc[[i]]
    #dfAloneAndRight.iloc[[f]]=HandRight.iloc[[i]]
    
    #dfAloneAndRight.iloc[[(i)]]=dfAlone.loc[[i]]
    
    #dfAloneAndRight.iloc[[2*i]]=lineLeft
    #dfAloneAndRight.iloc[[2*i+1]]=lineRight
    
    i=i+1
    #print(dfAloneAndRight)



In [ ]:
print(2*3)1

In [ ]:
dfAloneAndRight.head(15)

In [ ]:
#Now let's define the names of the columns
dfAloneAndRight.columns =['Time', ' Frame ID','Hand Type',' # hands', 'Position X', 'Position Y', 'Position Z', 'Velocity X',
             'Velocity Y', 'Velocity Z','Pitch','Roll','Yaw', 'Wrist Pos X', 'Wrist Pos Y', 'Wrist Pos Z', 'Elbow pos X',
             'Elbow Pos Y','Elbow Pos Z','Grab Strenth', 'Grab Angle', 'Pinch Strength']


In [ ]:
dfAloneAndRight.head(5)

In [ ]:
# Turnj each frame of 2 rows- Left hand and Right hand into one row:

In [ ]:
res = pd.DataFrame(columns=('lib', 'qty1', 'qty2'))
res


In [ ]:
HandRight.loc[[31197]]

In [ ]:
line = DataFrame({"onset": 30.0, "length": 1.3}, index=[3])
df2 = concat([df.iloc[:2], line, df.iloc[2:]]).reset_index(drop=True)

while len(dfHandRight)<len(dfAlone)
{
    
    RandomRows = HandRight.sample(n=7,replace=True)
    #and merge it with the HandRight df that already exist

    framesHandToAdd=[HandRight,RandomRows]

    NewRightHand = pd.concat(framesHandToAdd)
    HandRight=dfHandRight1.reset_index()
    HandRight.drop(['index'],axis='columns', inplace=True)
}